### DISCUSSION TITLE
How to see which apps have access to termux storage and revoke the permission?

### DISCUSSION THREAD

**Discussion by unoukujou** [termux/termux-app#2975] [Q&A] [answered]
On this page:

https://wiki.termux.com/wiki/Internal_and_external_storage

The part where it says "Access Termux from a file manager".

In the past, I noticed when accessing the Termux files from other apps, I had to give permission to the app in order for it to access Termux files. 

How to see all the apps that I gave permission to access Termux? And how to revoke the permission? I think all apps can access Termux files for some reason and I want to lockdown termux files to be used only by one app and no other app will be able to read/write there. Thanks.

**Comment by agnostic-apollo** [comment, marked as answer]
Run `adb shell dumpsys activity permissions` and check entries for `sourcePkg=com.termux` or check `/data/system/urigrants.xml` file with root.

https://cs.android.com/android/platform/superproject/+/master:frameworks/base/services/core/java/com/android/server/uri/UriGrantsManagerService.java;l=150

Currently, no api in termux to revoke access but might be added in next version. Reinstalling should likely work.

**Reply by unoukujou** [reply]
Cool thanks. Would be good to have a built in way to list all apps that have access and revoke permissions when needed. In the beginning I was only testing it out and now I'm stuck with too many apps that have access. Maybe I'll try to reinstall those apps to see if they get removed before I reinstall the whole termux.

**Comment by unoukujou** [comment]
Strange thing. I installed [Simple Text Editor](https://f-droid.org/en/packages/com.maxistar.textpad/) and I didn't give it any permissions. Yet it can freely use Termux storage. Just choosing File->Open from the Simple Text Editor menu, I can browse through Termux files and open any file I want. And this is a brand new installed app that I just installed for testing. I gave it no permission at all. All apps can just freely access the storage without permission? Or I screwed something up maybe?

**Reply by agnostic-apollo** [reply]
If it opens the android's document picker and you chose termux from side bar, you can browse through any termux home files, but it won't have access until you pick a file and it won't have access to other files. If document picker is opened in directory mode, then it will be able to access the directory you choose and any files underneath.

**Reply by agnostic-apollo** [reply]
>I gave it no permission at all.

I don't know what you else are you expecting to give it permission.

https://developer.android.com/guide/topics/providers/document-provider

**Reply by unoukujou** [reply]
So basically, because it opens the android picker, it can use any files from Termux. But do you know if only "me" the user can access files this way? Or apps can read files from termux storage without me knowing about it? 

Basically I was trying to use Termux storage to store more sensitive files that I only want one app to access these files. Therefore its more like a secure local storage rather than in the regular filesystem where apps can read files. You know what I mean? I just wanted to have a folder there to put some files that no other apps can read/write except one app which I permit.

I get confused because with some apps, if I try to access a termux folder, the document picker ask me if I want to allow this app to access the folder. But other apps like Simple Text Editor doest ask any permission, it just let's it have whatever file it wants.

**Reply by agnostic-apollo** [reply]
Apps don't have access unless you pick a file or directory with document picker.

If you pick termux home directory, app will be able to access all files under it. So best choose a sub directory if you want to limit access and only grant access to trustable apps or only grant file access/use file picker if app supports it.

**Reply by unoukujou** [reply]
Thanks. It makes more sense to me now. I'll make a sub directory in termux just for the app to store files there and I'll be sure no other apps will access it. Now just to clean up the mess that I made because I gave a lot of apps access to termux home dir from using this feature over the past few months. Maybe I'll just reinstall termux like you suggested before. Sounds to be the cleanest way to do it.

**Reply by agnostic-apollo** [reply]
If using github builds, adding `revokeDocumentPermission(BASE_DIR.getAbsolutePath());` at following and creating a build, installing apk and opening termux home in `Files` app should revoke access for home (not subdirectories), confirm with `dumpsys`. Install normal build back after that.

https://github.com/termux/termux-app/blob/211340781bcb0576ffe100fda2e0ed87d7333d53/app/src/main/java/com/termux/filepicker/TermuxDocumentsProvider.java#L90-91

**Reply by unoukujou** [reply]
It's a little bit over my head so I'll be more comfortable with a reinstall, plus the added benefit that I can start fresh and this time I'll be more careful with how I use the storage. And I don't really know what sub directories I gave to certain apps so I want to be sure I clear out everything and not just Home. I just didn't want to have to reinstall all pkg's and set everything up all over again. Termux is kind of a special app that really becomes more unique over time as you use it and set things up how you like it and make little changes here and there. So it's not so simple to reinstall and start all over. But I'd rather do that then leave random apps with access to the storage so its better off in the long run.

**Reply by amorphous** [reply]
This is a shame.

It seems odd to allow access to all termux files just by the touch of a button.

Surely (if I'm understanding you correctly) this is a security flaw???

**Reply by agnostic-apollo** [reply]
Pretty much all dangerous android permissions can be granted with a touch of a button that show in app's permissions list. And termux implements the android's standard SAF provider.

Note that if you know which apps have been granted access to a termux document uri with `dumpsys`, then  you can go in those apps App Info -> Storage and then click `Clear access`. That will remove all uri accesses granted to the app, whether for termux or any other app. There doesn't seem to be a android UI way to revoke only termux app uri's access.
